app/Platform/Actions/UpdateGameMetrics.php

```diff
@@ -81,23 +84,34 @@ public function execute(Game $game): void
         $tmp = $playersTotalChange;
         $tmp = $playersHardcoreChange;
 
-        // ad-hoc updates for player games, so they can be updated the next time a player updates their profile
-        // Note that those might be multiple thousand entries depending on a game's players count
-
-        $game->playerGames()
-            ->where(function ($query) use ($game) {
-                $query->whereNot('points_weighted_total', '=', $game->TotalTruePoints);
-            })
-            ->update([
-                'update_status' => 'weighted_points_outdated',
-                'points_weighted_total' => $game->TotalTruePoints,
-            ]);
-
-        $game->playerGames()
+        // Ad-hoc updates for player games metrics and player metrics after achievement set version changes
+        // Note: this might dispatch multiple thousands of jobs depending on a game's players count
+        $affectedPlayerGamesQuery = $game->playerGames()
             ->where(function ($query) use ($game) {
                 $query->whereNot('achievement_set_version_hash', '=', $game->achievement_set_version_hash)
                     ->orWhereNull('achievement_set_version_hash');
-            })
+            });
+
+        // add all affected player games to the update queue in batches
+        if (config('queue.default') !== 'sync') {
+            (clone $affectedPlayerGamesQuery)
+                ->whereNull('update_status')
+                ->orderByDesc('last_played_at')
+                ->chunk(1000, function (Collection $chunk) {
+                    // map and dispatch this chunk as a batch of jobs
+                    Bus::batch(
+                        $chunk->map(
+                            fn (PlayerGame $playerGame) => new UpdatePlayerGameMetricsJob($playerGame->user_id, $playerGame->game_id)
+                        )
+                    )
+                        ->onQueue('player-game-metrics-batch')
+                        ->dispatch();
+                });
```

Comment:
I'm a bit concerned that `UpdateGameMetrics` updates both the achievement set metrics (`achievements_published`, `achievements_unpublished`, `points_total`) and set player metrics (`players_total`, `players_hardcore`).

Because of this dual-purposing, it gets fired when the achievement set changes (fine) or when an achievement is unlocked (potentially very frequently for these common games):
https://github.com/RetroAchievements/RAWeb/blob/8d9eff006f1fafea8cd7f1977df54d1b808d0d8e/app/Platform/EventServiceProvider.php#L58-L61

If I understand the logic here, every time any user unlocks an achievement in Super Mario World, the entire Super Mario World userbase will be queued up to have their game metrics re-evaluated. Even if the filter on the hash changing prevents that, we shouldn't have to generate that hash and query the player_games table to see if it's been changed just when some random user unlocks an achievement that won't have any impact on anything other than possibly the total number of players in the set.

I think we should separate the player count stuff out of this action and put it in a separate one that doesn't generate the achievement_set_hash or attempt to batch up changes. Maybe `UpdateGamePlayerMetrics`?

Reply:
valid concerns, i had similar thoughts. it's very hard to tell how it scales without running it on the server. not all of the unlocks will cascade into a game metrics update - it can only ever be queued once.

do i understand correctly, `UpdateGamePlayerMetrics` would only include the affected player games batch updates after the achievement set version changed?
i think returning early after `GameMetricsUpdated::dispatch()` in `UpdateGameMetrics` when there was no achievement set version change would result in the same behaviour for now?

keeping it everything together makes it easier to reason about but there's definitely room for improvement. especially since the batching is very greedy in itself and could be much smarter by checking what actually changed. 
personally, i'd like to see it in action before doing too much optimization or splitting that might not be necessary.

Reply:
I'm just thinking that these really popular games are almost always going to have _someone_ playing them. As such, they're constantly going to be calculating the hash and scanning the 40k `player_games` rows looking for hash mismatches. Since we can pretty easily associate the hash to specific changes to the achievements (which happen very infrequently), we should avoid that overhead as much as possible.

> UpdateGamePlayerMetrics would only include the affected player games batch updates after the achievement set version changed?

Yes, leave this action for changes to the achievement set. And add a new action for changes to the players of the set. Changing the set would also update the players as the player count is dependent on the "exposed" achievements.  But the much more frequent unlocking of achievements (and the inverse - resetting of achievements) would be streamlined to just update the player counts.

It could be even further optimized to only be invoked on unlocking the first achievement or resetting the last achievement - though that doesn't play into the automatic mapping of the unlock event to the metrics update through the listener.

> i think returning early after GameMetricsUpdated::dispatch() in UpdateGameMetrics when there was no achievement set version change would result in the same behaviour for now?

That would avoid scanning the 40k rows, but it would still have to generate the hash every time an achievement was unlocked in the set just to see if it changed.

> not all of the unlocks will cascade into a game metrics update - it can only ever be queued once.

Yes, but if a large number of people are playing a popular game at the same time and unlocking a reasonable number of achievements, the game metrics update will be reinserted back into the queue almost as quickly as it cycles out.

As you've said, it would be best to see exactly how things perform in production, but I don't think we can simulate the load the site actually sees.